Feat: Add Undo and Redo functionality

ravjot07 opened the pull request:
Add: Undo and Redo for pen stoke..

https://github.com/user-attachments/assets/fd655522-fb6a-4ae5-a74a-810a7d90daed

quozl commented:
Thanks.

* [ ] your commit messages do not follow our guidance, in particular
  * [ ] "added" should be "add",
  * [ ] the first lines are too long,
  * [ ] there's nothing but first lines,
  * [ ] "implemented" is vague,
  * [ ] typo "stoke",
* [ ] you have missing newline on files,
* [ ] the undo and redo icons do not follow Sugar style, please use our existing icons for undo and redo; remember each activity is just one of several, so don't focus only on this repository.

@walterbender, please review?  (You posted on issue #14).

ravjot07 commented:
> Thanks.
> 
> * [ ]  your commit messages do not follow our guidance, in particular
>   
>   * [ ]  "added" should be "add",
>   * [ ]  the first lines are too long,
>   * [ ]  there's nothing but first lines,
>   * [ ]  "implemented" is vague,
>   * [ ]  typo "stoke",
> * [ ]  you have missing newline on files,
> * [ ]  the undo and redo icons do not follow Sugar style, please use our existing icons for undo and redo; remember each activity is just one of several, so don't focus only on this repository.
> 
> @walterbender, please review? (You posted on issue #14).

@quozl can you review this undo and redo button styles and i have resolved most of the suggestions as you mentioned sir

quozl commented:
You can find edit-undo.svg in sugar-artwork repository, icons/scalable/actions

ravjot07 commented:
> sugar-artwork

@quozl Undo and Redo button Updated according to sugar-artwork

quozl commented:
Thanks.  @chimosky, @walterbender any review comments?

chimosky commented on css/activity.css:
You should look at the diffs of your changes before committing them as this will help you see the no new line at EOF changes that were made, although the one here doesn't seem like it was added because of your changes.

ravjot07 replied:
Noted, I’ve resolved this issue...

chimosky commented on icons/redo-button.svg:
Sugar has it's own redo and undo icons, you should use them, you can find them in [sugar-artwork](https://github.com/sugarlabs/sugar-artwork/tree/master/icons/scalable/actions).

ravjot07 replied:
i have added svgs from sugar artwork in later pr

chimosky commented on js/activity.js:
```diff
@@ -323,4 +401,4 @@ define(function (require) {
   // Export the DrawingApp class
   return DrawingApp;
-});
+});
```
Another no new line at EOF.

chimosky reviewed (commented):
Your commit message says "Implemented a stoke history data structure", I'm guessing you meant "stroke history".

If that's the case then you should edit the commit message.

chimosky commented:
Reviewed, not tested.

I'm genuinely curious why you add lots of spaces in most of your commits, a commit message says one thing and after looking at the commit most of the changes are new line additions which is just noise.

ravjot07 commented:
> Reviewed, not tested.
> 
> I'm genuinely curios why you add lots of spaces in most of your commits, a commit message says one thing and after looking at the commit most of the changes are new line additions which is just noise.

@chimosky This could be due to the linting tools configured in my setup, which might be automatically adjusting whitespace or adding new lines. I’ll review my configuration to ensure cleaner commits going forward.

ravjot07 commented:
@quozl @chimosky

ravjot07 commented:
@quozl if there are no further changes we should merge this pr

quozl commented:
https://github.com/sugarlabs/connect-the-dots-activity/pull/26#issuecomment-2526357186 is my review of commit messages, no changes have happened  yet.

ravjot07 committed:
feat: add Undo and Redo functionality
- Implement stroke history data structure for tracking changes
- Add Undo and Redo buttons to the main toolbar
- Implement undo and redo methods and integrate them with the UI

Signed-off-by: ravjot07 <ravu2004@gmail.com>

ravjot07 commented:
Updated in PR #38

quozl commented:
No reason to close, please push your changes here.

ravjot07 commented:
@quozl done, this feature is working now we can merge this one...
